### PR TITLE
Filters

### PR DESCRIPTION
**Pull request by jpcima**
#30

This is the multi-mode filter.
It has been renamed and moved into the `sfz` namespace.

I have also added the generated code, which is an important amount of code.
To make the PR better, I isolated it in its own commit.

### CODE REVIEW

**Comment by jpcima** [comment]
Just a fix for an error just noticed.
Copying the wrong size in case of filter set to none.

```
-                std::memcpy(ch_out, ch_in, NCh * nframes * sizeof(float));
+                std::memcpy(ch_out, ch_in, nframes * sizeof(float));
```

File: src/sfizz/SfzFilter.cpp [hunk 1]
```diff
@@ -0,0 +1,256 @@
...
+}
+
+template <unsigned NCh>
+template <class F>
+void Filter<NCh>::Impl::process(F &filter, const float *const in[NCh], float *const out[NCh], float cutoff, float q, float pksh, unsigned nframes)
+{
+    filter.setCutoff(cutoff);
+    filter.setQ(q);
+    filter.setPkShGain(pksh);
+    filter.compute(nframes, const_cast<float **>(in), const_cast<float **>(out));
```
Comment (paulfd) [inline src/sfizz/SfzFilter.cpp:125]: I'm not a super big fan of the const casts... Is there a reason Faust generates the interfaces as `float**`?

File: src/sfizz/SfzFilter.h [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+ */
+template <unsigned NCh>
+class Filter {
+public:
+    Filter();
+    ~Filter();
+
+    /**
+       Set up the filter constants.
+       Run it exactly once after instantiating.
```
Comment (paulfd) [inline src/sfizz/SfzFilter.h:31]: Is it true? It seems to me this could be changed live.

File: src/sfizz/SfzFilter.h [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+    void clear();
+
+    /**
+       Process one cycle of the filter without modulating cutoff or Q.
+       `cutoff` is a frequency expressed in Hz.
+       `q` is a resonance expressed in dB.
+       `pksh` is a peak/shelf gain expressed in dB.
+       `in[i]` and `out[i]` may refer to identical buffers, for in-place processing
+     */
+    void process(const float *const in[NCh], float *const out[NCh], float cutoff, float q, float pksh, unsigned nframes);
```
Comment (paulfd) [inline src/sfizz/SfzFilter.h:47]: Would you be OK using AudioSpans and Spans in these interfaces? To make it more consistent with the rest of the codebase. I can translate them.

**Review by paulfd — APPROVED** [review]
For the rest I think it is OK for now, we can tweak the interface while hooking the code to the rest of the codebase. Thanks!

File: src/sfizz/SfzFilter.cpp (hunk 1 above)
Comment (jpcima) [reply src/sfizz/SfzFilter.cpp:125]: I'm afraid it's the way it is for a long time, unfortunately.
I don't see for it any good reason, it's a thing to deal with.

File: src/sfizz/SfzFilter.h (hunk 2 above)
Comment (jpcima) [reply src/sfizz/SfzFilter.h:31]: This sets up only constants dependent on the sample rate.
The sample rate is not changed live. (or is it?)
Run it multiple times then, it should be fine.

File: src/sfizz/SfzFilter.h (hunk 3 above)
Comment (jpcima) [reply src/sfizz/SfzFilter.h:47]: On examination, I can't say I love AudioSpan particularly.

It fattens significantly the calling protocol.
It has always 2 pointers (MaxChannels), and 2 longs per instance.

---

`processModulated` Mono
- `raw` : 5 pointers, 1 int
- `AudioSpan` : 10 pointers, 10 longs

---

`processModulated` Stereo
- `raw` : 7 pointers, 1 int
- `AudioSpan` : 10 pointers, 10 longs

---

As I see, it's far of what I would call "0-cost" feature, and my point of view is rather not use this extra expressivity, which requires on such cost.

I think especially of ARM machines which benefits of light call frames, for register-only passing.

In bodies of functions, then yes maybe, but not as part of the function's declaration.

Comment (paulfd) [reply src/sfizz/SfzFilter.h:47]: There's slightly less than this and it's true that for such big interfaces the cost of possibly redundant lengths add up. As I said we can discuss it later anyway and I am interested about the ARM argument 🙂

**Pull request merged**